src/riak_kv_token_manager.erl

```diff
@@ -0,0 +1,610 @@
+                        {none, UpdQueues} ->
+                            {noreply,
+                                State#state{
+                                    associations = UpdAssocs,
+                                    queues = UpdQueues,
+                                    grants =
+                                        maps:remove(
+                                            TokenID,
+                                            State#state.grants
+                                        )
+                                }
+                            };
+                        {{NextSession, NextVerifyList}, UpdQueues}
+                                when NextVerifyList == VerifyList ->
+                            %% Only requests with the same VerifyList should be
+                            %% queued
+                            UpdGrants =
+                                maps:put(
+                                    TokenID,
+                                    {local, NextSession, VerifyList, 0},
+                                    State#state.grants
+                                ),
+                            NextSession ! granted,
+                            %% Check that the downstream nodes are still aware
+                            %% of this grant (in case, for example, they have
+                            %% restarted in between)
+                            ok =
+                                downstream_renew(
+                                    VerifyList,
+                                    TokenID,
+                                    NextSession
+                                ),
+                            {noreply,
+                                State#state{
+                                    associations = UpdAssocs,
+                                    queues = UpdQueues,
+                                    grants = UpdGrants
+                                }
+                            }
+                    end;
+                {upstream, _Manager, Pid} ->
+                    %% It is possible that this is a remote PID that has failed
+                    %% No need to promote form queue in this case, just remove
+                    {noreply,
+                        State#state{
+                            associations = UpdAssocs,
+                            queues = Queues,
+                            grants =
+                                maps:remove(
+                                    TokenID,
+                                    State#state.grants
+                                )
+                        }
+                    };
+                _ ->
+                    %% The session may have been queued and not had a grant
+                    {noreply,
+                        State#state{associations = UpdAssocs, queues = Queues}
+                    }
+            end;
+        _ ->
+            %% No assocition is assumed to be a token manager
```

Comment:
```suggestion
            %% No association is assumed to be a token manager
```

Reply:
Now checks Ref/Pid is that of a monitored remote manager